### PR TITLE
Move XC course questions to subdirectory structure

### LINKED ISSUE #2209: MatplotlibDeprecationWarning in demoRandomPlot
The exampleCourse (XC 101) question `demoRandomPlot` throws a `MatplotlibDeprecationWarning`, which triggers creation of an issue every time it's accessed. It would be great to fix the plotting code in this question as described in the warning message:

```
<string>:20: MatplotlibDeprecationWarning: Adding an axes using the same arguments
as a previous axes currently reuses the earlier instance.  In a future version, a
new instance will always be created and returned.  Meanwhile, this warning can be
suppressed, and the future behavior ensured, by passing a unique label to each axes
instance.
```

### LINKED ISSUE #2571: New XC Question structure
With https://github.com/PrairieLearn/PrairieLearn/pull/2489 by @nicknytko merged, we now proceed onto manipulating the XC question directory structure.

- documentationQuestions
   - Code found in PrairieLearn’s readthedocs
- elementQuestions
   - Examples of all question features… Acts as a “unit test”
- demoQuestions
   - practical questions that might appear with varying difficulty.
- template questions
   - ideal starting points for new users

### PR DESCRIPTION
**Pull request by coatless**
In this PR, we:

- incorporate subdirectories into the XC course for better organization.
- truly fixed randomPlot's axis issue.

Specifically, we added **3** high-level directories:

- demo
   - demonstration questions, not the best
- element 
   - unit test questions / element gallery of options.
- template
   - re-usable questions

Underneath `demo` we have subcategories for:

- `autograder` and `autograder/python`
- `custom`
- `drawing`
- `manualGrade` 

Subdirectories are now:

```bash
tree -d exampleCourse/questions
├── demo
│   ├── autograder
│   │   ├── ansiOutput
│   │   ├── codeEditor
│   │   │   └── tests
│   │   ├── codeUpload
│   │   │   ├── submitted_code_samples
│   │   │   │   ├── fib_error_binary_file
│   │   │   │   ├── fib_error_fill_disk
│   │   │   │   ├── fib_error_fork_bomb
│   │   │   │   ├── fib_error_huge_stderr
│   │   │   │   ├── fib_error_huge_stdout
│   │   │   │   ├── fib_error_large_text_file
│   │   │   │   ├── fib_error_timeout_list
│   │   │   │   ├── fib_error_timeout_numpy
│   │   │   │   ├── fib_error_use_network
│   │   │   │   ├── fib_success_score_partial
│   │   │   │   ├── fib_success_score_perfect
│   │   │   │   └── fib_success_score_zero
│   │   │   └── tests
│   │   └── python
│   │       ├── numpy
│   │       │   └── tests
│   │       ├── pandas
│   │       │   └── tests
│   │       ├── plots
│   │       │   └── tests
│   │       ├── random
│   │       │   └── tests
│   │       └── square
│   │           └── tests
│   ├── calculation
│   ├── custom
│   │   ├── element
│   │   ├── gradeFunction
│   │   └── stylingScripts
│   │       └── clientFilesQuestion
│   ├── drawing
│   │   ├── centroid
│   │   ├── collarRod
│   │   ├── gradeVector
│   │   ├── graphs
│   │   ├── inclinedPlane
│   │   ├── liftingMechanism
│   │   ├── pulley
│   │   ├── simpleTutorial
│   │   └── vmDiagrams
│   ├── fixedCheckbox
│   │   └── clientFilesQuestion
│   ├── manualGrade
│   │   └── codeUpload
│   ├── markdownEditorLivePreview
│   ├── matrixAlgebra
│   ├── matrixComplexAlgebra
│   ├── overlayTriangle
│   ├── randomCheckbox
│   ├── randomDataFrame
│   ├── randomMultipleChoice
│   │   └── clientFilesQuestion
│   ├── randomPlot
│   ├── randomSymbolic
│   ├── studentFaces
│   └── studentNames
├── element
│   ├── checkbox
│   ├── code
│   ├── codeDocumentation
│   │   └── clientFilesQuestion
│   ├── drawingGallery
│   ├── fileDownload
│   │   └── clientFilesQuestion
│   ├── fileEditor
│   ├── graph
│   ├── integerInput
│   ├── markdown
│   ├── matrixComponentInput
│   ├── matrixLatex
│   ├── multipleChoice
│   ├── numberInput
│   ├── overlay
│   │   └── clientFilesQuestion
│   ├── panels
│   ├── prairieDrawFigure
│   ├── pythonVariable
│   ├── stringInput
│   ├── symbolicInput
│   ├── threeJS
│   │   └── clientFilesQuestion
│   └── variableOutput
└── template
    ├── checkbox
    ├── integerInput
    ├── integerInputWithDynamicFigure
    ├── multipleChoiceFromList
    ├── multipleChoiceMultipleCorrect
    ├── multipleChoiceNumeric
    ├── multipleChoiceSingleCorrect
    ├── multipleChoiceTrueFalseFromList
    ├── multipleChoiceWithFigure
    │   └── clientFilesQuestion
    ├── numberInput
    ├── numberInputGradingMethod
    └── numberInputTips
```

Close #2571 
Close #2209

----

Some _R_ scripts for the modification. Though, I'm sure there are some bash scripts I wrote last time.

```r
prefixes = c("demo", "element", "template")


## Mass rename ----
rename_problem_dir = function(folder_prefix = "demo") {
    
    dir_list = list.dirs(folder_prefix, 
                         full.names = FALSE, recursive = FALSE)
    
    
    file.rename(file.path(folder_prefix, dir_list), 
                file.path(folder_prefix, problem_name))
    
}

## Mass assessment data change ----
rename_problem_assessment = function(name_prefix = "demo") {
    assessment_files = list.files(pattern = ".json", recursive = T)
    
    for(i in seq_along(assessment_files)) {
        m = readLines(assessment_files[i])
        
        replace_m = gsub(paste0('{"id": "(',name_prefix,')([[:upper:]])(.*)}'), 
                         replacement = paste0('{"id": "', name_prefix, '/\\L\\2\\E\\3}'),
                         m,
                         perl = TRUE)
        
        writeLines(replace_m, assessment_files[i])
    }
}

## Modify question folders ----
# Change working directory
setwd("/Users/ronin/Desktop/PrairieLearn/exampleCourse/questions")

# Rock'n'Roll
sapply(
    c("demo", "element", "template"),
    rename_problem_dir
    )

## Modify question metadata ----

# Switch directory
setwd("/Users/ronin/Desktop/PrairieLearn/exampleCourse/courseInstances/Sp15")

# Change assessmentoverview
sapply(
    c("demo", "element", "template"),
    rename_problem_assessment
)
```

### CODE REVIEW

**Comment by coatless** [comment]
Not sure why moving all XC questions into subfolders is triggering issues with the course browser unit tests here:

https://github.com/PrairieLearn/PrairieLearn/runs/741058377#step:9:13404

**Comment by coatless** [comment]
46e1f0b addressed the test suite errors. They arose due to a hardcoded QID in `courseEditor.js`

**Comment by nicknytko** [comment]
Some others I found while `grep`'ing through the source:

```
docs/elements.md:1598:[element/element/codeDocumentation]: https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/elementElementCodeDocumentation
docs/pl-drawing/index.md:1206:The example [`QID: demoDrawingGradeVector`](https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/drawingGradeVector) illustrates all the `pl-drawing-button`
docs/python-grader/index.md:172:One can hook into library functions in the setup code to disallow students from accessing certain functions.  This example is taken from the `demoAutograderNumpy` question.

**Comment by coatless** [comment]
@mwest1066 merged in master. The subdir change should be good to go.

**Comment by coatless** [comment]
@nicknytko promise not to touch any XC course files for 24 hours? ;)

**Comment by coatless** [comment]
@nicknytko + @mwest1066 bumpity bump bump

File: docs/python-grader/index.md [hunk 1]
```diff
@@ -170,7 +170,7 @@ Be careful not to switch the ordering of the student and reference arguments.  T
 
 ## Banning/Disallowing library functions
 
-One can hook into library functions in the setup code to disallow students from accessing certain functions.  This example is taken from the `demoAutograderNumpy` question.
+One can hook into library functions in the setup code to disallow students from accessing certain functions.  This example is taken from the [demo/autograderNumpy] question.
```
Comment (nicknytko) [inline docs/python-grader/index.md:173]:
```suggestion
One can hook into library functions in the setup code to disallow students from accessing certain functions.  This example is taken from the [demo/autograder/python/numpy] question.
```

File: docs/python-grader/index.md [hunk 2]
```diff
@@ -194,3 +194,7 @@ The grading Python scripts will load any sensitive files into memory (setup code
 After grading, results will be written to the secret filename generated above.  If this file does not exist or the filename does not match then the grading job will fail and students will not receive points.  This is mostly a failsafe in case the grader code were to crash, but in theory could also prevent crafty students from writing their own results file.
 
 The grading job will now drop back to `root` in the `run.sh` script and will copy any output to the correct location, as expected by the external grading framework.
+
+
+
+[demo/autograderNumpy]: https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/autograderNumpy
```
Comment (nicknytko) [inline docs/python-grader/index.md:200]:
```suggestion
[demo/autograder/python/umpy]: https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/autograder/python/numpy
```

**Comment by coatless** [comment]
@nicknytko thanks! Fixed.

**Comment by coatless** [comment]
@mwest1066 saturday bump for review

**Comment by mwest1066** [comment]
Looks great! My only gripe is that the default ordering in the Questions list is weird, because it orders by Topic.

Could you fix the topics to match (at least roughly) the folder layout? I'm thinking something like all the `element/` questions would be in `Topic: Element`, the `demo/autograder` questions would be in `Topic: Autograder`, etc. It's fine if this is not a one-to-one match, but at the moment it's all over the place.

**Pull request merged**